src/v/cluster/partition_properties_stm.cc

```diff
@@ -0,0 +1,326 @@
+  storage::kvstore& kvstore,
+  config::binding<std::chrono::milliseconds> sync_timeout)
+  : raft::persisted_stm<raft::kvstore_backed_stm_snapshot>(
+      partition_properties_stm_snapshot, logger, raft, kvstore)
+  , _sync_timeout(std::move(sync_timeout))
+  , _state_snapshots({state_snapshot{
+      .writes_disabled = writes_disabled::no,
+      .update_offset = model::offset{}}}) {}
+
+ss::future<iobuf> partition_properties_stm::take_snapshot(model::offset o) {
+    if (o < _raft->start_offset()) {
+        throw std::invalid_argument(fmt::format(
+          "can not take raft snapshot at offset {} which is smaller than raft "
+          "start offset",
+          o,
+          _raft->start_offset()));
+    }
+    vlog(
+      _log.trace,
+      "taking snapshot at offset: {}, state snapshots: {}",
+      o,
+      fmt::join(_state_snapshots, ", "));
+
+    auto it = std::upper_bound(
+      _state_snapshots.begin(),
+      _state_snapshots.end(),
+      o,
+      [](model::offset o, const state_snapshot& current_update) {
+          return o < current_update.update_offset;
+      });
+
+    vassert(
+      it != _state_snapshots.begin(),
+      "Partition properties state machine holds an invariant that first "
+      "snapshot in state snapshots field is smaller than raft start offset "
+      "therefore the situation in which first element of the "
+      "_state_snapshots is returned is a critical failure. Requested "
+      "offset {}, state snapshots: {}",
+      o,
+      fmt::join(_state_snapshots, ", "));
+
+    // move iterator one step back to obtain cached snapshot with smaller
+    // offset
+    --it;
+    vlog(
+      _log.trace,
+      "returning indexed snapshot with offset: {}, requested offset: {}",
+      it->update_offset,
+      o);
+    co_return serde::to_iobuf(
+      raft_snapshot{.writes_disabled = it->writes_disabled});
+}
+
+ss::future<> partition_properties_stm::apply_local_snapshot(
+  raft::stm_snapshot_header header, iobuf&& buffer) {
+    vlog(_log.debug, "Applying local snapshot with offset {}", header.offset);
+    auto snapshot = serde::from_iobuf<local_snapshot>(std::move(buffer));
+    _state_snapshots = std::move(snapshot.state_updates);
+    co_return;
+}
+
+ss::future<raft::stm_snapshot> partition_properties_stm::take_local_snapshot(
+  ssx::semaphore_units apply_units) {
+    vlog(_log.debug, "Taking local snapshot");
+
+    auto last_applied = last_applied_offset();
+
+    // cleanup snapshot index
+    // delete all obsolete snapshot indicies
+    auto it = std::upper_bound(
+      _state_snapshots.begin(),
+      _state_snapshots.end(),
+      _raft->start_offset(),
+      [](model::offset o, const state_snapshot& current_update) {
+          return o < current_update.update_offset;
+      });
+    vassert(
+      it != _state_snapshots.begin(),
+      "State snapshot vector must always contain element that is strictly "
+      "smaller than the underlying log start offset. Snapshot vector {}",
```

Comment:
it may, however, be exactly equal to it here, as `std::upper_bound` gives us the one that is strictly bigger